spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+        self.pid: Optional[int] = None
+        self.last_keepalive_time: Optional[int] = None
+        self.running_os_cmd_index: Optional[int] = None
+        self.rc: Optional[int] = None
+        self.console_output: Optional[str] = None
+
+    def change_state(self, new_state: str):
+        assert new_state in self.state.transitions, (
+            "Illegal transition from " + self.state.name + " to " + new_state + "."
+        )
+        self.state = _ParallelCommandInfo.state_dict[new_state]
+
+    @property
+    def state_repr(self) -> str:
+        state_str = self.state.name
+        if state_str == "running" and self.running_os_cmd_index is not None:
+            number_of_os_cmds = len(self.cmd["script"])
+            state_str = (
+                f"{state_str} ({self.running_os_cmd_index + 1}/{number_of_os_cmds})"
+            )
+        elif state_str in ("failed", "terminated") and os.name == "nt":
+            state_str = "failed/terminated"
+        # we know ANSI commands are available because otherwise
+        # the status table would not be being displayed in the first place
+        return color(state_str, self.state.display_color)
+
+
+def project_run_parallel_group(
+    project_dir: Path,
+    cmd_names: List[str],
+    *,
+    overrides: Dict[str, Any] = SimpleFrozenDict(),
+    force: bool = False,
+    dry: bool = False,
+) -> None:
+    """Run a parallel group of commands. Note that because of the challenges of managing
+    parallel console output streams it is not possible to specify a value for 'capture' as
+    when executing commands serially. Essentially, with parallel groups 'capture==True'.
+
+    project_dir (Path): Path to project directory.
+    cmd_names: the names of the spaCy projects commands to run.
+    overrides (Dict[str, Any]): Optional config overrides.
+    force (bool): Force re-running, even if nothing changed.
+    dry (bool): Perform a dry run and don't execute commands.
+    """
+    config = load_project_config(project_dir, overrides=overrides)
+    commands = {cmd["name"]: cmd for cmd in config.get("commands", [])}
+    check_spacy_commit = check_bool_env_var(ENV_VARS.PROJECT_USE_GIT_VERSION)
+    parallel_group_status_queue = Manager().Queue()
+    max_parallel_processes = config.get("max_parallel_processes")
+    cmd_infos = [
+        _ParallelCommandInfo(cmd_name, commands[cmd_name], cmd_index)
+        for cmd_index, cmd_name in enumerate(cmd_names)
+    ]
+
+    with working_dir(project_dir) as current_dir:
+        for cmd_info in cmd_infos:
+            check_deps(cmd_info.cmd, cmd_info.cmd_name, project_dir, dry)
+            if (
+                not check_rerun(
+                    current_dir, cmd_info.cmd, check_spacy_commit=check_spacy_commit
+                )
+                and not force
+            ):
+                cmd_info.change_state("not rerun")
+        rmtree(PARALLEL_LOGGING_DIR_NAME, ignore_errors=True)
+        os.mkdir(PARALLEL_LOGGING_DIR_NAME)
```

Comment:
Also - I might have missed it, but do we check whether this directory has been created by the user before we delete it? In general I'm in favor for using `tempfile.TemporaryDirectory()` for something like this, since (a) the output is printed to the console anyway and (b) if users want to keep the log files, they can copy them to a non-temporary location afterwards.

Reply:
Forgot about (c), which is also important: it automatically avoids all conflicts with pre-existing file/directory names.

Reply:
I think you have some good points there - having default path(s) is certainly useful, and it's rather unlikely users have a `parrTmp` directory. Regarding the last point: a temporary directory can be configured to not be deleted once the corresponding context has been closed.

In general I'd go for the standard way nonetheless, as I think the advantages don't really justify the non-standard approach here. I don't mind this approach either though as long as we are checking and warning (with confirmation maybe) that the existing `parrTmp` directory will be deleted. Maybe a second opinion would be useful for this.

Reply:
I'd rather go the temporary directory route, but:

1) I don't want to delete the temporary directory after the parallel group has completed in case a user wants to access the files
2) I don't want temporary directories to be created with each invocation of a workflow and never deleted

Perhaps a good compromise would be to delete the temporary directory if everything succeeded and retain it if something went wrong? What do you think?

Reply:
1. AFAIK on Linux temporary files and directories are deleted after a certain number of days or on system restart and in Windows never by default.
2. This is probably a dealbreaker then for temporary directories, see 1.

I think logs should be kept independently from success or failure. Perhaps we do have a convention regarding temporary files already though?

Reply:
@rmitsch and I discussed this offline and decided to go for standard temporary directories; I have changed the code accordingly.